[BEAM-2732] Metrics rely on statesampler state

pabloem opened the pull request:
This PR depends on PR #4375.

This PR is a backward-compatible change that allows metrics to rely on statesampler state to provide the current container.

This change should enable all clients to stop tracking metrics with its own context, and start using statesampler for that.

Benchmarks showed no impact on performance. Benchmark done using an ML Criteo pipeline that processes 10gb.

Results Before Change (5 runs):
Mean: 1809.2s, Stdev: 227.1622768

Results With Change (5 runs):
Mean: 1758.4s, Stdev: 93.46282684

pabloem commented:
Run Python PostCommit

pabloem commented:
Run Python PostCommit

pabloem commented:
Run Python PostCommit

pabloem commented:
Run Python PostCommit

pabloem commented:
r: @robertwb 
Here's the goal for the statesampler changes. To have everything rely on the same functionality to track context. This change is backwards compatible, but I'll make sure to port all clients immediately, so the code won't remain in this state for more than a couple weeks.

pabloem commented:
* There's also the issue of step names for the operation. Right now I'm simply passing step_name in, which seems safe, but I'm not sure.

robertwb commented on sdks/python/apache_beam/transforms/util.py:
```diff
@@ -221,6 +220,10 @@ def __init__(self,
     self._ignore_next_timing = False
+
+    # TODO(pabloem) - This should rely on internal counters. Not user metrics.
+    from apache_beam.metrics import Metrics
```
Why this change?

robertwb commented on sdks/python/apache_beam/transforms/util.py:
```diff
@@ -221,6 +220,10 @@ def __init__(self,
     self._data = []
     self._ignore_next_timing = False
+
+    # TODO(pabloem) - This should rely on internal counters. Not user metrics.
```
No, this is a user-level transform.

robertwb commented on sdks/python/apache_beam/metrics/execution.py:
```diff
@@ -135,18 +136,40 @@ def set_metrics_supported(self, supported):
+  def current_container(self):
+    """Returns the current MetricsContainer."""
+    sampler = statesampler.EXECUTION_STATE_SAMPLERS.current_sampler()
+    if sampler is None:
```
When would this be false? Can we simply assert it'll be present after your CL rather than keeping both the old and new paths in parallel?

robertwb commented on sdks/python/apache_beam/runners/worker/statesampler.py:
```diff
@@ -49,18 +67,43 @@ def __init__(self, prefix, counter_factory,
 
+    # TODO(pabloem) - Remove this once all clients register the
+    # sampler independently.
+    self.register()
```
As I mentioned before, I'd really like to avoid being in this state, even temporarily.

robertwb commented on sdks/python/apache_beam/runners/worker/statesampler.py:
```diff
@@ -49,18 +67,43 @@ def __init__(self, prefix, counter_factory,
+  def create_and_register_tracker(
+      prefix, counter_factory, sampling_period_ms=DEFAULT_SAMPLING_PERIOD_MS):
+    sampler = StateSampler(prefix, counter_factory, sampling_period_ms)
+    sampler.register()
```
Won't this be a double registration?

robertwb commented on sdks/python/apache_beam/runners/worker/statesampler_slow.py:
```diff
@@ -49,19 +55,15 @@ def start(self):
     self.name = name
     self.counter = counter
     self.nsecs = 0
+    self.metrics_container = metrics_container
```
Adding a counter container here seems out-of-place, and who knows what else might get added.

Instead, I'd suggest an architecture where the there is the ability to get the local state (which is an object that can be used as a key and has basic information like step name) which other frameworks (e.g. metrics) consume in a self-contained way. This also reduces the scope of the global state we must store to access these contexts implicitly from user code.

robertwb replied:
I agree that tagging state here may be useful for cleanup when the state is no longer needed, so I suppose this is OK.

robertwb commented on sdks/python/apache_beam/runners/worker/statesampler.py:
```diff
@@ -29,6 +31,22 @@
+    return self._current_sampler
+
+  def set_sampler(self, sampler):
+    self._current_sampler = sampler
```
Throw an error if the sampler was already set.

robertwb commented on sdks/python/apache_beam/runners/worker/statesampler.py:
```diff
@@ -29,6 +31,22 @@
+    self._current_sampler = sampler
+
+
+EXECUTION_STATE_SAMPLERS = ExecutionStateSamplers()
```
Perhaps the name should indicate that this is per-thread? In general, creating a global singleton class rather than module-level functions is a Javaism.

robertwb commented on sdks/python/apache_beam/runners/worker/statesampler.py:
```diff
@@ -29,6 +31,22 @@
+    self._current_sampler = None
+
+  def current_sampler(self):
+    return self._current_sampler
```
Throw an error if it was not set?

robertwb commented on sdks/python/apache_beam/runners/worker/statesampler.py:
```diff
@@ -49,18 +67,43 @@ def __init__(self, prefix, counter_factory,
+
+  def register(self):
+    EXECUTION_STATE_SAMPLERS.set_sampler(self)
+    self._registered = True
```
Is this value very queried?

robertwb commented on sdks/python/apache_beam/runners/worker/statesampler.py:
```diff
@@ -49,18 +67,43 @@ def __init__(self, prefix, counter_factory,
+    sampler.register()
+    return sampler
+
+  def register(self):
```
Maybe register_on_current_thread? I'd actually have the caller call EXECUTION_STATE_SAMPLERS.set_sampler directly, so the interaction of getting the sampler is symmetrc with setting it.

pabloem commented:
Let me know if PR 4504 seems like a more sound implementation, or if PR 4387 (this one) seems more sensible.

pabloem commented:
Changes from the current commit:
1. Addet get/set_current_tracker methods
2. Removed unnecessary ExecutionStateTrackers class.
3. Removed register() call from StateSampler constructor. Instead, a state sampler registers itself when start() is called (similar to java), or a user can use set_current_tracker
4. Removed simple_tracker, and make_tracker calls. StateSampler should be created via its constructor.

Regarding your questions:
1. The difference between a counter and a metric cell? 
- A metric cell manages locking and commit state of metrics.
- Counters created by a CounterFactory (I believe you were referring to those?) are used internally only.

2. Is the thread-local variable accessed for every metric update?
Yes, that is the current state (currently, and it will not change in this PR). Definitely an optimization we could consider.

pabloem commented:
Rebased

pabloem commented:
Run Python PostCommit

pabloem commented:
@robertwb PTAL

pabloem commented:
Run Python PostCommit

pabloem commented:
Run Python PostCommit

pabloem commented:
Run Python PostCommit

pabloem commented:
Run Python PostCommit

pabloem commented:
Run Python PostCommit

pabloem commented:
Run Python PostCommit

pabloem commented:
Run Python PostCommit

pabloem commented:
All integration tests but `test_streaming_wordcount_it` are passing. I'll try to fix that.....

pabloem commented:
Run Python PostCommit

pabloem commented:
Run Python PostCommit

pabloem commented:
Python Postcommit tests are passing. I'm not sure why the `mvn clean install -pl sdks/python -am -am...` are broken, as they are a subset of the Postcommit suite:

![image](https://user-images.githubusercontent.com/1301740/38341543-00e4e738-382e-11e8-9b24-35e45dffbc10.png)

@robertwb PTAL

pabloem commented:
Rebased change.

pabloem commented:
Run Python PostCommit

pabloem commented:
@robertwb PTAL

pabloem commented:
Rebased change.

pabloem commented:
Run Python PostCommit.

pabloem commented:
Retest this please

pabloem commented:
Retest this please.

pabloem commented:
Run Python PostCommit

robertwb commented on sdks/python/apache_beam/runners/direct/direct_runner.py:
```diff
@@ -338,6 +337,7 @@ def run_pipeline(self, pipeline):
     from apache_beam.runners.direct.transform_evaluator import \
       TransformEvaluatorRegistry
     from apache_beam.testing.test_stream import TestStream
+    from apache_beam.metrics.execution import MetricsEnvironment
```
Undo this change? Metrics should not be changed to depend on anything in the runners package. (Either that or the lazy import should be made there.)

robertwb replied:
Or perhaps don't import metrics.execution when importing the public (write) Metrics api.

pabloem replied:
That is reasonable, and I agree. I've fixed this by having a lazy import in the metrics execution module.
Currently, we're working under the idea that state sampler is the global context provider, so metrics rely on it. I'd think that makes it into a module that provides a service to other modules that require/handle context management. Though I'd agree that a more loose coupling would be good to have.

robertwb commented on sdks/python/apache_beam/runners/direct/executor.py:
```diff
@@ -290,70 +293,87 @@ def __init__(self, transform_evaluator_registry, evaluation_context,
 
-    with scoped_metrics_container:
-      evaluator.start_bundle()
+    evaluator.start_bundle()
```
Why is this not called from within start_state?

pabloem replied:
The function `attempt_call` is already called within start_state (because side input initialization happens before `attempt_call` is called. I've documented this in the `attempt_call` docstring, but we could do it differently if you think that'd make more sense.

robertwb replied:
Yes, I think it'd be cleaner to have all states explicitly entered within attempt_call than make this an implicit assumption about the caller.

pabloem replied:
Done.

robertwb commented on sdks/python/apache_beam/runners/worker/operations.py:
```diff
@@ -133,24 +133,25 @@ def __init__(self, name_context, spec, counter_factory, state_sampler):
     self.metrics_container = MetricsContainer(self.name_context.metrics_name())
-    self.scoped_metrics_container = ScopedMetricsContainer(
-        self.metrics_container)
+    self.scoped_metrics_container = ScopedMetricsContainer()
```
Can this just be deleted?

pabloem replied:
Right. Done.

pabloem replied:
Ah nevermind. The op.scoped_metrics_container is used by Dataflow worker code. Needs to be available.

robertwb replied:
OK. Please add a comment (with a JIRA) here about that.

pabloem replied:
Done. Thanks Robert.

robertwb commented on sdks/python/apache_beam/transforms/util.py:
```diff
@@ -221,6 +220,9 @@ def __init__(self,
     self._data = []
     self._ignore_next_timing = False
+
+    from apache_beam.metrics import Metrics
```
Undo this change. Transforms should be free to use metrics (at least the public metrics write API).

pabloem replied:
Done.

pabloem commented:
Thanks Robert!

pabloem commented:
Run Python PostCommit

pabloem commented:
Retest this please.

pabloem commented:
Squashed commits. Letting tests run.

robertwb commented:
Jenkins: retest this please

pabloem commented:
PreCommits seem to be broken, also on my machine after I rebased.

robertwb commented:
What errors were you getting on your machine?

pabloem commented:
Monsterpaste:

```
======================================================================
ERROR: test_basics_with_type_check (apache_beam.examples.cookbook.group_with_coder_test.GroupWithCoderTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/examples/cookbook/group_with_coder_test.py", line 53, in test_basics_with_type_check
    '--output=%s.result' % temp_path])
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/examples/cookbook/group_with_coder.py", line 118, in run
    | WriteToText(known_args.output))
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/io/textio.py", line 522, in __init__
    skip_header_lines=skip_header_lines)
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/io/textio.py", line 117, in __init__
    validate=validate)
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/io/filebasedsource.py", line 119, in __init__
    self._validate()
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/options/value_provider.py", line 124, in _f
    return fnc(self, *args, **kwargs)
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/io/filebasedsource.py", line 176, in _validate
    match_result = FileSystems.match([pattern], limits=[1])[0]
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/io/filesystems.py", line 166, in match
    return filesystem.match(patterns, limits)
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/io/filesystem.py", line 600, in match
    raise BeamIOError("Match operation failed", exceptions)
BeamIOError: Match operation failed with exceptions {'/tmp/tmpJcX3Kr*': BeamIOError("List operation failed with exceptions {'/tmp': OSError(2, 'No such file or directory')}",)}
-------------------- >> begin captured logging << --------------------
root: INFO: Missing pipeline option (runner). Executing pipeline using the default runner: DirectRunner.
--------------------- >> end captured logging << ---------------------

======================================================================
ERROR: test_basics_without_type_check (apache_beam.examples.cookbook.group_with_coder_test.GroupWithCoderTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/examples/cookbook/group_with_coder_test.py", line 74, in test_basics_without_type_check
    '--output=%s.result' % temp_path])
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/examples/cookbook/group_with_coder.py", line 118, in run
    | WriteToText(known_args.output))
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/io/textio.py", line 522, in __init__
    skip_header_lines=skip_header_lines)
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/io/textio.py", line 117, in __init__
    validate=validate)
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/io/filebasedsource.py", line 119, in __init__
    self._validate()
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/options/value_provider.py", line 124, in _f
    return fnc(self, *args, **kwargs)
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/io/filebasedsource.py", line 176, in _validate
    match_result = FileSystems.match([pattern], limits=[1])[0]
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/io/filesystems.py", line 166, in match
    return filesystem.match(patterns, limits)
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/io/filesystem.py", line 600, in match
    raise BeamIOError("Match operation failed", exceptions)
BeamIOError: Match operation failed with exceptions {'/tmp/tmpisQlPp*': BeamIOError("List operation failed with exceptions {'/tmp': OSError(2, 'No such file or directory')}",)}
-------------------- >> begin captured logging << --------------------
root: INFO: Missing pipeline option (runner). Executing pipeline using the default runner: DirectRunner.
--------------------- >> end captured logging << ---------------------

======================================================================
ERROR: test_multiple_output_pardo (apache_beam.examples.cookbook.multiple_output_pardo_test.MultipleOutputParDo)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/examples/cookbook/multiple_output_pardo_test.py", line 56, in test_multiple_output_pardo
    '--output=%s' % result_prefix])
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/examples/cookbook/multiple_output_pardo.py", line 154, in run
    lines = p | ReadFromText(known_args.input)
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/io/textio.py", line 522, in __init__
    skip_header_lines=skip_header_lines)
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/io/textio.py", line 117, in __init__
    validate=validate)
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/io/filebasedsource.py", line 119, in __init__
    self._validate()
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/options/value_provider.py", line 124, in _f
    return fnc(self, *args, **kwargs)
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/io/filebasedsource.py", line 176, in _validate
    match_result = FileSystems.match([pattern], limits=[1])[0]
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/io/filesystems.py", line 166, in match
    return filesystem.match(patterns, limits)
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/io/filesystem.py", line 600, in match
    raise BeamIOError("Match operation failed", exceptions)
BeamIOError: Match operation failed with exceptions {'/tmp/tmpnmwepq*': BeamIOError("List operation failed with exceptions {'/tmp': OSError(2, 'No such file or directory')}",)}
-------------------- >> begin captured logging << --------------------
root: INFO: Missing pipeline option (runner). Executing pipeline using the default runner: DirectRunner.
--------------------- >> end captured logging << ---------------------

======================================================================
ERROR: test_basics (apache_beam.examples.wordcount_debugging_test.WordCountTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/examples/wordcount_debugging_test.py", line 52, in test_basics
    '--output=%s.result' % temp_path])
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/examples/wordcount_debugging.py", line 134, in run
    | 'FilterText' >> beam.ParDo(FilterTextFn('Flourish|stomach')))
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/io/textio.py", line 522, in __init__
    skip_header_lines=skip_header_lines)
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/io/textio.py", line 117, in __init__
    validate=validate)
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/io/filebasedsource.py", line 119, in __init__
    self._validate()
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/options/value_provider.py", line 124, in _f
    return fnc(self, *args, **kwargs)
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/io/filebasedsource.py", line 176, in _validate
    match_result = FileSystems.match([pattern], limits=[1])[0]
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/io/filesystems.py", line 166, in match
    return filesystem.match(patterns, limits)
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/io/filesystem.py", line 600, in match
    raise BeamIOError("Match operation failed", exceptions)
BeamIOError: Match operation failed with exceptions {'/tmp/tmpSmPq3X*': BeamIOError("List operation failed with exceptions {'/tmp': OSError(2, 'No such file or directory')}",)}
-------------------- >> begin captured logging << --------------------
root: INFO: Missing pipeline option (runner). Executing pipeline using the default runner: DirectRunner.
--------------------- >> end captured logging << ---------------------

======================================================================
ERROR: test_basics (apache_beam.examples.wordcount_minimal_test.WordCountMinimalTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/examples/wordcount_minimal_test.py", line 47, in test_basics
    '--output=%s.result' % temp_path])
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/examples/wordcount_minimal.py", line 100, in run
    lines = p | ReadFromText(known_args.input)
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/io/textio.py", line 522, in __init__
    skip_header_lines=skip_header_lines)
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/io/textio.py", line 117, in __init__
    validate=validate)
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/io/filebasedsource.py", line 119, in __init__
    self._validate()
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/options/value_provider.py", line 124, in _f
    return fnc(self, *args, **kwargs)
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/io/filebasedsource.py", line 176, in _validate
    match_result = FileSystems.match([pattern], limits=[1])[0]
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/io/filesystems.py", line 166, in match
    return filesystem.match(patterns, limits)
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/io/filesystem.py", line 600, in match
    raise BeamIOError("Match operation failed", exceptions)
BeamIOError: Match operation failed with exceptions {'/tmp/tmprNRvNX*': BeamIOError("List operation failed with exceptions {'/tmp': OSError(2, 'No such file or directory')}",)}

======================================================================
ERROR: test_basics (apache_beam.examples.wordcount_test.WordCountTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/examples/wordcount_test.py", line 46, in test_basics
    '--output=%s.result' % temp_path])
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/examples/wordcount.py", line 90, in run
    lines = p | 'read' >> ReadFromText(known_args.input)
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/io/textio.py", line 522, in __init__
    skip_header_lines=skip_header_lines)
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/io/textio.py", line 117, in __init__
    validate=validate)
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/io/filebasedsource.py", line 119, in __init__
    self._validate()
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/options/value_provider.py", line 124, in _f
    return fnc(self, *args, **kwargs)
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/io/filebasedsource.py", line 176, in _validate
    match_result = FileSystems.match([pattern], limits=[1])[0]
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/io/filesystems.py", line 166, in match
    return filesystem.match(patterns, limits)
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/io/filesystem.py", line 600, in match
    raise BeamIOError("Match operation failed", exceptions)
BeamIOError: Match operation failed with exceptions {'/tmp/tmpnAxeZ2*': BeamIOError("List operation failed with exceptions {'/tmp': OSError(2, 'No such file or directory')}",)}
-------------------- >> begin captured logging << --------------------
root: INFO: Missing pipeline option (runner). Executing pipeline using the default runner: DirectRunner.
--------------------- >> end captured logging << ---------------------

======================================================================
ERROR: test_sink_transform (apache_beam.io.avroio_test.TestAvro)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/io/avroio_test.py", line 390, in test_sink_transform
    readback = p | avroio.ReadFromAvro(path + '*') | beam.Map(json.dumps)
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/io/avroio.py", line 137, in __init__
    self._source = _AvroSource(file_pattern, min_bundle_size, validate=validate)
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/io/filebasedsource.py", line 119, in __init__
    self._validate()
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/options/value_provider.py", line 124, in _f
    return fnc(self, *args, **kwargs)
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/io/filebasedsource.py", line 176, in _validate
    match_result = FileSystems.match([pattern], limits=[1])[0]
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/io/filesystems.py", line 166, in match
    return filesystem.match(patterns, limits)
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/io/filesystem.py", line 600, in match
    raise BeamIOError("Match operation failed", exceptions)
BeamIOError: Match operation failed with exceptions {'/tmp/tmpN4Vb7G*': BeamIOError("List operation failed with exceptions {'/tmp': OSError(2, 'No such file or directory')}",)}

```

aaltay commented:
https://github.com/apache/beam/pull/5167 should address those errors.

pabloem commented:
Run Python PostCommit

pabloem commented:
I believe this is ready to be merged.
